### ISSUE TITLE
CherryPy fails to build

### ISSUE THREAD

**Issue by sjau** [NixOS/nixpkgs#91220]
**Describe the bug**
CherryPy fails to build.

**To Reproduce**
```
==================================== ERRORS ====================================                                                      

http_server = <generator object http_server.<locals>.start_srv at 0x7ffff4342040>                                                     
unix_sock_file = '\x00cheroot-test-sockete0501561-9e84-4684-86ce-c38972c61513'                                                        

    @pytest.fixture                                                                                                                   
    def peercreds_enabled_server_and_client(http_server, unix_sock_file):                                                             
        """Construct a test server with `peercreds_enabled`."""                                                                       
        httpserver = http_server.send(unix_sock_file)                                                                                 
        httpserver.gateway = _TestGateway                                                                                             
        httpserver.peercreds_enabled = True                                                                                           
>       return httpserver, get_server_client(httpserver)                                                                              

http_server = <generator object http_server.<locals>.start_srv at 0x7ffff4342040>                                                     
httpserver = <cheroot.server.HTTPServer object at 0x7ffff43960a0>                                                                     
unix_sock_file = '\x00cheroot-test-sockete0501561-9e84-4684-86ce-c38972c61513'                                                        

cheroot/test/test_server.py:175:                                                                                                      
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _                                                       
cheroot/testing.py:153: in get_server_client                                                                                          
    return _TestClient(server)                                                                                                        
        server     = <cheroot.server.HTTPServer object at 0x7ffff43960a0>                                                             
cheroot/testing.py:84: in __init__                                                                                                    
    self._http_connection = self.get_connection()                                                                                     
        self       = <cheroot.testing._TestClient object at 0x7ffff439e460>                                                           
        server     = <cheroot.server.HTTPServer object at 0x7ffff43960a0>                                                             
cheroot/testing.py:96: in get_connection                                                                                              
    return conn_cls(name)                                                                                                             
        conn_cls   = <class 'http.client.HTTPConnection'>                                                                             
        name       = '\x00cheroot-test-sockete0501561-9e84-4684-86ce-c38972c61513:0'                                                  
        self       = <cheroot.testing._TestClient object at 0x7ffff439e460>                                                           
/nix/store/n0b076p351ma864q38if4yglsg99hw2s-python3-3.8.3/lib/python3.8/http/client.py:831: in __init__                               
    self._validate_host(self.host)                                                                                                    
        blocksize  = 8192                                                                                                             
        host       = '\x00cheroot-test-sockete0501561-9e84-4684-86ce-c38972c61513:0'                                                  
        port       = None                                                                                                             
        source_address = None                                                                                                         
        timeout    = <object object at 0x7ffff71d5c90>                                                                                
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _                                                       

self = <http.client.HTTPConnection object at 0x7ffff43a6a00>                                                                          
host = '\x00cheroot-test-sockete0501561-9e84-4684-86ce-c38972c61513'                                                                  

    def _validate_host(self, host):                                                                                                   
        """Validate a host so it doesn't contain control characters."""                                                               
        # Prevent CVE-2019-18348.                                                                                                     
        match = _contains_disallowed_url_pchar_re.search(host)                                                                        
        if match:                                                                                                                     
>           raise InvalidURL(f"URL can't contain control characters. {host!r} "                                                       
                             f"(found at least {match.group()!r})")                                                                   
E           http.client.InvalidURL: URL can't contain control characters. '\x00cheroot-test-sockete0501561-9e84-4684-86ce-c38972c61513' (found at least '\x00')

host       = '\x00cheroot-test-sockete0501561-9e84-4684-86ce-c38972c61513'                                                            
match      = <re.Match object; span=(0, 1), match='\x00'>                                                                             
self       = <http.client.HTTPConnection object at 0x7ffff43a6a00>                                                                    

/nix/store/n0b076p351ma864q38if4yglsg99hw2s-python3-3.8.3/lib/python3.8/http/client.py:1193: InvalidURL                               
_______ ERROR at setup of test_peercreds_unix_sock_with_lookup[abstract] _______

http_server = <generator object http_server.<locals>.start_srv at 0x7ffff42327b0>                                                     
unix_sock_file = '\x00cheroot-test-socket4100880e-acaa-4078-aa3b-5200244d0c44'                                                        

    @pytest.fixture                                                                                                                   
    def peercreds_enabled_server_and_client(http_server, unix_sock_file):                                                             
        """Construct a test server with `peercreds_enabled`."""                                                                       
        httpserver = http_server.send(unix_sock_file)                                                                                 
        httpserver.gateway = _TestGateway                                                                                             
        httpserver.peercreds_enabled = True                                                                                           
>       return httpserver, get_server_client(httpserver)                                                                              

http_server = <generator object http_server.<locals>.start_srv at 0x7ffff42327b0>                                                     
httpserver = <cheroot.server.HTTPServer object at 0x7ffff41ad310>                                                                     
unix_sock_file = '\x00cheroot-test-socket4100880e-acaa-4078-aa3b-5200244d0c44'                                                        

cheroot/test/test_server.py:175:                                                                                                      
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _                                                       
cheroot/testing.py:153: in get_server_client                                                                                          
    return _TestClient(server)                                                                                                        
        server     = <cheroot.server.HTTPServer object at 0x7ffff41ad310>                                                             
cheroot/testing.py:84: in __init__                                                                                                    
    self._http_connection = self.get_connection()                                                                                     
        self       = <cheroot.testing._TestClient object at 0x7ffff41ada00>                                                           
        server     = <cheroot.server.HTTPServer object at 0x7ffff41ad310>                                                             
cheroot/testing.py:96: in get_connection                                                                                              
    return conn_cls(name)                                                                                                             
        conn_cls   = <class 'http.client.HTTPConnection'>                                                                             
        name       = '\x00cheroot-test-socket4100880e-acaa-4078-aa3b-5200244d0c44:0'                                                  
        self       = <cheroot.testing._TestClient object at 0x7ffff41ada00>                                                           
/nix/store/n0b076p351ma864q38if4yglsg99hw2s-python3-3.8.3/lib/python3.8/http/client.py:831: in __init__                               
    self._validate_host(self.host)                                                                                                    
        blocksize  = 8192                                                                                                             
        host       = '\x00cheroot-test-socket4100880e-acaa-4078-aa3b-5200244d0c44:0'                                                  
        port       = None                                                                                                             
        source_address = None                                                                                                         
        timeout    = <object object at 0x7ffff71d5c90>                                                                                
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _                                                       

self = <http.client.HTTPConnection object at 0x7ffff41adc70>                                                                          
host = '\x00cheroot-test-socket4100880e-acaa-4078-aa3b-5200244d0c44'                                                                  

    def _validate_host(self, host):                                                                                                   
        """Validate a host so it doesn't contain control characters."""                                                               
        if match:                                                                                                                     
                             f"(found at least {match.group()!r})")                                                                   
E           http.client.InvalidURL: URL can't contain control characters. '\x00cheroot-test-socket4100880e-acaa-4078-aa3b-5200244d0c44' (found at least '\x00')
host       = '\x00cheroot-test-socket4100880e-acaa-4078-aa3b-5200244d0c44'                                                            
=============================== warnings summary ===============================
    _issue_warning_captured(deprecated.JUNIT_XML_DEFAULT_FAMILY, config.hook, 2)                                                      
cheroot/__main__.py                 3      3     0%   3-6                                                                             
cheroot/test/test_server.py       126      4    97%   60, 160, 186, 219                                                               
cheroot/test/test_ssl.py          178     55    69%   86, 227-354, 411, 459, 463                                                      
cheroot/test/webtest.py           323    193    40%   56-59, 64-66, 74-81, 96-98, 127-128, 132-136, 147-152, 161, 165, 173, 198-228, 238-246, 251-296, 301, 305-310, 318-323, 327-338, 342-351, 355-362, 366-371, 375-381, 385-392, 396-401, 405-410, 414-419, 439, 451-454, 466-481, 499, 523-526, 530, 539, 545, 567-570, 596-605                                                                               
cheroot/testing.py                 78     11    86%   48-49, 74-75, 124-128, 132, 144-146                                             
cheroot/workers/threadpool.py     152     54    64%   25, 28, 122, 132-136, 138-139, 183, 188-189, 202-207, 211-222, 225-228, 234-250, 261-262, 291-292, 297-308, 323                                                                                                       
cheroot/wsgi.py                   162     46    72%   92, 96, 147, 160, 170-173, 179, 183, 203, 205, 215, 282-292, 316, 332-349, 353-356, 360-364, 418-424                                                                                                                  
-------------------------------------------------------------                                                                         
TOTAL                            3746   1119    70%                                                                                   

6 files skipped due to complete coverage.                                                                                             
========================== slowest 10 test durations ===========================                                                      
4.11s call     cheroot/test/test_conn.py::test_HTTP11_Timeout_after_request                                                           
3.41s call     cheroot/test/test_conn.py::test_keepalive_conn_management                                                              
2.00s call     cheroot/test/test_conn.py::test_HTTP11_Timeout[True]                                                                   
0.50s call     cheroot/test/test_conn.py::test_HTTP11_pipelining                                                                      
  Server does not correctly read trailers/ending of the previous HTTP request, thus the second request fails as the server tries to pa====== 77 passed, 30 deselected, 2 xfailed, 1 warning, 2 errors in 21.19s ======
f_null.o bf_null.c                                                                                                                    
./../../torrent/data/block_transfer.h:132:78: warning: throw will always call terminate() [-Wterminate]                               
  132 |     throw internal_error("BlockTransfer::~BlockTransfer() peer_info not NULL");                                               
./../../torrent/data/block_transfer.h:132:78: note: in C++11 destructors default to noexcept                                          
gcc -I.. -I../.. -I../modes -I../asn1 -I../evp -I../../include  -fPIC -DOPENSSL_PIC -DOPENSSL_THREADS -D_REENTRANT -DDSO_DLFCN -DHAVE_HA1_ASM -DSHA256_ASM -DSHA512_ASM -DMD5_ASM -DAES_ASM -DVPAES_ASM -DBSAES_ASM -DWHIRLPOOL_ASM -DGHASH_ASM -DECP_NISTZ256_ASM   -c -o bcheroot/test/test_makefile.py      31      4    87%   27-30                                                                                                                                                                                                                    
cheroot/test/test_server.py       126      4    97%   60, 160, 186, 219                                                                                                                                                                                                        
cheroot/test/test_ssl.py          178     55    69%   86, 227-354, 411, 459, 463                                                                                                                                                                                               
cheroot/test/webtest.py           323    193    40%   56-59, 64-66, 74-81, 96-98, 127-128, 132-136, 147-152, 161, 165, 173, 198-228, 238-246, 251-296, 301, 305-310, 318-323, 327-338, 342-351, 355-362, 366-371, 375-381, 385-392, 396-401, 405-410, 414-419, 439, 451-454, 466-481, 499, 523-526, 530, 539, 545, 567-570, 596-605                                                                                                                                                                                                                           
cheroot/testing.py                 78     11    86%   48-49, 74-75, 124-128, 132, 144-146                                                                                                                                                                                      
cheroot/workers/threadpool.py     152     54    64%   25, 28, 122, 132-136, 138-139, 183, 188-189, 202-207, 211-222, 225-228, 234-250, 261-262, 291-292, 297-308, 323                                                                                                          
cheroot/wsgi.py                   162     46    72%   92, 96, 147, 160, 170-173, 179, 183, 203, 205, 215, 282-292, 316, 332-349, 353-356, 360-364, 418-424                                                                                                                     
-------------------------------------------------------------                                                                                                                                                                                                                  
TOTAL                            3746   1119    70%                                                                                                                                                                                                                            

6 files skipped due to complete coverage.                                                                                                                                                                                                                                      
Coverage XML written to file coverage.xml                                                                                                                                                                                                                                      

========================== slowest 10 test durations ===========================                                                                                                                                                                                               
4.11s call     cheroot/test/test_conn.py::test_HTTP11_Timeout_after_request                                                                                                                                                                                                    
3.41s call     cheroot/test/test_conn.py::test_keepalive_conn_management                                                                                                                                                                                                       
2.00s call     cheroot/test/test_conn.py::test_HTTP11_Timeout[True]                                                                                                                                                                                                            
2.00s call     cheroot/test/test_conn.py::test_HTTP11_Timeout[False]                                                                                                                                                                                                           
0.61s call     cheroot/test/test_conn.py::test_598                                                                                                                                                                                                                             
0.50s call     cheroot/test/test_server.py::test_stop_interrupts_serve                                                                                                                                                                                                         
0.50s call     cheroot/test/test_conn.py::test_HTTP11_pipelining                                                                                                                                                                                                               
0.30s call     cheroot/test/test_ssl.py::test_ssl_adapters[builtin]                                                                                                                                                                                                            
0.20s call     cheroot/test/test_conn.py::test_HTTP11_persistent_connections                                                                                                                                                                                                   
0.20s call     cheroot/test/test_conn.py::test_streaming_11[False]                                                                                                                                                                                                             
=========================== short test summary info ============================                                                                                                                                                                                               
XFAIL cheroot/test/test_conn.py::test_Chunked_Encoding                                                                                                                                                                                                                         
  Server does not correctly read trailers/ending of the previous HTTP request, thus the second request fails as the server tries to parse b'Content-Type: application/json\r\n' as a Request-Line. This results in HTTP status code 400, instead of 413Ref: https://github.com/cherrypy/cheroot/issues/69                                                                                                                                                                                                                                                     
XFAIL cheroot/test/test_conn.py::test_598                                                                                                                                                                                                                                      
  Sometimes this test fails due to low timeout. Ref: https://github.com/cherrypy/cherrypy/issues/598                                                                                                                                                                           
====== 77 passed, 30 deselected, 2 xfailed, 1 warning, 2 errors in 21.19s ======
gcc -I.. -I../.. -I../modes -I../asn1 -I../evp -I../../include  -fPIC -DOPENSSL_PIC -DOPENSSL_THREADS -D_REENTRANT -DDSO_DLFCN -DHAVE_DLFCN_H -m64 -DL_ENDIAN -O3 -Wall -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DRC4_ASM -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DMD5_ASM -DAES_ASM -DVPAES_ASM -DBSAES_ASM -DWHIRLPOOL_ASM -DGHASH_ASM -DECP_NISTZ256_ASM   -c -o bf_null.o bf_null.c                                                                                                                          
In file included from ./../data/block.h:42,                                                                                                                                                                                                                                    
                 from peer.cc:39:                                                                                                                                                                                                                                              
./../../torrent/data/block_transfer.h: In destructor 'torrent::BlockTransfer::~BlockTransfer()':                                                                                                                                                                               
./../../torrent/data/block_transfer.h:129:74: warning: throw will always call terminate() [-Wterminate]                                                                                                                                                                        
  129 |     throw internal_error("BlockTransfer::~BlockTransfer() block not NULL");                                                                                                                                                                                            
      |                                                                          ^
./../../torrent/data/block_transfer.h:129:74: note: in C++11 destructors default to noexcept                                                                                                                                                                                   
./../../torrent/data/block_transfer.h:132:78: warning: throw will always call terminate() [-Wterminate]                                                                                                                                                                        
  132 |     throw internal_error("BlockTransfer::~BlockTransfer() peer_info not NULL");                                                                                                                                                                                        
      |                                                                              ^
cheroot/test/test_makefile.py      31      4    87%   27-30
cheroot/test/test_server.py       126      4    97%   60, 160, 186, 219
cheroot/test/test_ssl.py          178     55    69%   86, 227-354, 411, 459, 463
cheroot/test/webtest.py           323    193    40%   56-59, 64-66, 74-81, 96-98, 127-128, 132-136, 147-152, 161, 165, 173, 198-228, 238-246, 251-296, 301, 305-310, 318-323, 327-338, 342-351, 355-362, 366-371, 375-381, 385-392, 396-401, 405-410, 414-419, 439, 451-454, 466-481, 499, 523-526, 530, 539, 545, 567-570, 596-605
cheroot/testing.py                 78     11    86%   48-49, 74-75, 124-128, 132, 144-146
cheroot/workers/threadpool.py     152     54    64%   25, 28, 122, 132-136, 138-139, 183, 188-189, 202-207, 211-222, 225-228, 234-250, 261-262, 291-292, 297-308, 323
cheroot/wsgi.py                   162     46    72%   92, 96, 147, 160, 170-173, 179, 183, 203, 205, 215, 282-292, 316, 332-349, 353-356, 360-364, 418-424
-------------------------------------------------------------
TOTAL                            3746   1119    70%

6 files skipped due to complete coverage.
Coverage XML written to file coverage.xml

========================== slowest 10 test durations ===========================
4.11s call     cheroot/test/test_conn.py::test_HTTP11_Timeout_after_request
3.41s call     cheroot/test/test_conn.py::test_keepalive_conn_management
2.00s call     cheroot/test/test_conn.py::test_HTTP11_Timeout[True]
2.00s call     cheroot/test/test_conn.py::test_HTTP11_Timeout[False]
0.61s call     cheroot/test/test_conn.py::test_598
0.50s call     cheroot/test/test_server.py::test_stop_interrupts_serve
0.50s call     cheroot/test/test_conn.py::test_HTTP11_pipelining
0.30s call     cheroot/test/test_ssl.py::test_ssl_adapters[builtin]
0.20s call     cheroot/test/test_conn.py::test_HTTP11_persistent_connections
0.20s call     cheroot/test/test_conn.py::test_streaming_11[False]
=========================== short test summary info ============================
XFAIL cheroot/test/test_conn.py::test_Chunked_Encoding
  Server does not correctly read trailers/ending of the previous HTTP request, thus the second request fails as the server tries to parse b'Content-Type: application/json\r\n' as a Request-Line. This results in HTTP status code 400, instead of 413Ref: https://github.com/cherrypy/cheroot/issues/69
XFAIL cheroot/test/test_conn.py::test_598
  Sometimes this test fails due to low timeout. Ref: https://github.com/cherrypy/cherrypy/issues/598
====== 77 passed, 30 deselected, 2 xfailed, 1 warning, 2 errors in 21.19s ======
gcc -I.. -I../.. -I../modes -I../asn1 -I../evp -I../../include  -fPIC -DOPENSSL_PIC -DOPENSSL_THREADS -D_REENTRANT -DDSO_DLFCN -DHAVE_DLFCN_H -m64 -DL_ENDIAN -O3 -Wall -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DRC4_ASM -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DMD5_ASM -DAES_ASM -DVPAES_ASM -DBSAES_ASM -DWHIRLPOOL_ASM -DGHASH_ASM -DECP_NISTZ256_ASM   -c -o bf_null.o bf_null.c
In file included from ./../data/block.h:42,
                 from peer.cc:39:
./../../torrent/data/block_transfer.h: In destructor 'torrent::BlockTransfer::~BlockTransfer()':
./../../torrent/data/block_transfer.h:129:74: warning: throw will always call terminate() [-Wterminate]
  129 |     throw internal_error("BlockTransfer::~BlockTransfer() block not NULL");
      |                                                                          ^
./../../torrent/data/block_transfer.h:129:74: note: in C++11 destructors default to noexcept
./../../torrent/data/block_transfer.h:132:78: warning: throw will always call terminate() [-Wterminate]
  132 |     throw internal_error("BlockTransfer::~BlockTransfer() peer_info not NULL");
      |                                                                              ^
./../../torrent/data/block_transfer.h:132:78: note: in C++11 destructors default to noexcept
gcc -I.. -I../.. -I../modes -I../asn1 -I../evp -I../../include  -fPIC -DOPENSSL_PIC -DOPENSSL_THREADS -D_REENTRANT -DDSO_DLFCN -DHAVE_DLFCN_H -m64 -DL_ENDIAN -O3 -Wall -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DRC4_ASM -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DMD5_ASM -DAES_ASM -DVPAES_ASM -DBSAES_ASM -DWHIRLPOOL_ASM -DGHASH_ASM -DECP_NISTZ256_ASM   -c -o bf_buff.o bf_buff.c
In file included from ./../../globals.h:41,
                 from ./../../download/download_main.h:43,
                 from peer.cc:41:
../../../rak/priority_queue_default.h: In destructor 'rak::priority_item::~priority_item()':
../../../rak/priority_queue_default.h:56:95: warning: throw will always call terminate() [-Wterminate]
   56 |       throw torrent::internal_error("priority_item::~priority_item() called on a queued item.");
      |                                                                                               ^
../../../rak/priority_queue_default.h:56:95: note: in C++11 destructors default to noexcept
builder for '/nix/store/6j2clxcqd047raz7z4an329d7vinqypr-python3.8-cheroot-8.2.1.drv' failed with exit code 1
cannot build derivation '/nix/store/n97vcq82irw22via2qhm19szksgsfzq1-python3.8-cherrypy-18.3.0.drv': 1 dependencies couldn't be built
```

**Notify maintainers**
@unode 

**Metadata**
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.43, NixOS, 20.09pre228204.467ce5a9f45 (Nightingale)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.5`
 - channels(hyper): `""`
 - channels(root): `"nixos-20.09pre230335.9480bae3370"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

**Comment by unode** [comment]
This is a cheroot problem and still present in the latest release (`8.3.0`).
It looks like an upstream issue, possibly a python 3.8 incompatibility.

Tagging also @jonringer @B4dM4n

**Comment by unode** [comment]
Also reported upstream: https://github.com/cherrypy/cheroot/issues/297

**Comment by unode** [comment]
On our end, we can workaround the failure by disabling the two `[abstract]` tests.

Any objections from other maintainers?

**Comment by jonringer** [comment]
Main issue is that the "http" base library no longer allows control characters in the url. I just disabled the tests in my PR, but upstream should probably change/remove the tests to react to the base libraries changing behavior.

**Comment by webknjaz** [comment]
Fixed in v8.3.1
